src/Exercise/PhpPromotion.php

```diff
@@ -0,0 +1,163 @@
+                $notPromoted,
+                implode('" & "', $notPromoted)
+            ));
+        }
+
+        if ($promoted->get('config')) {
+            return Failure::fromNameAndReason($this->getName(), 'Property "config" should not be promoted');
+        }
+
+        $expectedProperties = [
+            'basePath' => 'protected',
+            'key' => 'private',
+            'visitor' => 'private',
+        ];
+
+        $properties = collect($reflectionClass->getProperties());
+        $properties = $properties->flatMap(fn (\ReflectionProperty $prop) => [
+            $prop->getName() => $this->getPropertyVisibility($prop)
+        ])->getArrayCopy();
+        ksort($properties);
+
+        if ($changedVisibility = array_keys(array_diff_assoc($expectedProperties, $properties))) {
+            return Failure::fromNameAndReason($this->getName(), pluralise(
+                'Visibility changed for property "%s"',
+                $changedVisibility,
+                implode('" & "', $changedVisibility)
+            ));
+        }
+
+        $types = $properties->map(fn (\ReflectionProperty $prop) => $prop->getType()?->getName());
+        $expected = ['visitor' => 'Closure', 'key' => 'string', 'basePath' => 'string'];
+        if ([] !== $typeDiff = array_diff_assoc($expected, $types->getArrayCopy())) {
+            return Failure::fromNameAndReason($this->getName(), pluralise(
+                'Property "%s" should not have changed type',
+                array_keys($typeDiff),
+                implode('" & "', array_keys($typeDiff))
+            ));
+        }
+
+        // Test Invocation
+        $args = [
+            'visitor' => \Closure::fromCallable(fn (array $line) => $line[2]),
+            'key' => 'value',
+            'config' => ['basePath' => '/some/base/path']
+        ];
+
+        $obj = new $className(...$args);
+
+        $actual = $properties
+            ->each(fn (\ReflectionProperty $prop) => $prop->setAccessible(true))
+            ->map(fn (\ReflectionProperty $prop) => $prop->isInitialized($obj) ? $prop->getValue($obj) : null)
+            ->getArrayCopy();
+
+        $dataFailures = array_filter([
+            'visitor' => $args['visitor'] !== $actual['visitor'],
+            'key' => $args['key'] !== $actual['key'],
+            'basePath' => $args['config']['basePath'] !== $actual['basePath'],
+        ]);
+
+        if (count($dataFailures)) {
+            return Failure::fromNameAndReason($this->getName(), pluralise(
+                'Data not correctly set on property "%s"',
+                array_keys($dataFailures),
+                implode('" & "', array_keys($dataFailures))
+            ));
+        }
+
+        return new Success($this->getName());
+    }
+
+    private function getPropertyVisibility(\ReflectionProperty $prop): string
+    {
+        return match (true) {
+            $prop->isPrivate() => 'private',
+            $prop->isProtected() => 'protected',
+        default => 'public',
```

Comment:
alignment

Reply:
I thought this to, but not according to phpcs and phpstorm 🤷‍♂️

Reply:
wut